mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -335,6 +349,105 @@ def find_mindmeld_annotation_re_matches(text):
+        # MindMeld Entities are not present in text
+        if len(matches) == 0:
+            unann_spans = [(0, len(text))]
+            return unann_spans
+        for match in matches:
+            entity_start, entity_end = match.span()
+            entity_text = match.group(1)
+
+            unann_spans.append((prev_entity_end, entity_start))
+            entity_text_start = entity_start + 1
+            unann_spans.append(
+                (entity_text_start, entity_text_start + len(entity_text))
+            )
+            prev_entity_end = entity_end
+
+        if len(unann_spans) > 0:
+            unann_spans.append((prev_entity_end, len(text)))
+
+        # Filter out spans that have a length of 0
+        unann_spans = [span for span in unann_spans if span[1] - span[0] > 0]
+        return unann_spans
+
+    @staticmethod
+    def unann_to_ann_idx_map(unann_spans):
+        """Create a vector mapping indexes from the unannotated text to the original
+        text.
+
+        Args:
+            unann_spans (List[Tuple(int, int)]): The list of spans where each span
+                is a section of the original text excluding annotation markup. The
+                first element of the tuple is the start index and the second is the
+                ending index + 1.
+        Returns:
+            unann_to_ann_idx_map (List[Tuple(int, int)]): A vector where the value at
+                each index represents the mapping of the position of a single character
+                in the unannotated text to the position in the original text.
+        """
+        unann_to_ann_idx_map = []
+        for unann_span in unann_spans:
+            start, end = unann_span
+            for i in range(start, end):
+                unann_to_ann_idx_map.append(i)
+        return unann_to_ann_idx_map
+
+    @staticmethod
+    def convert_token_idx_unann_to_ann(tokens, unann_to_ann_idx_map):
+        """In-place function that reverts the token start indices to the
+        index of the character in the orginal text with annotations.
+
+        Args:
+            unann_to_ann_idx_map (List[Tuple(int, int)]): A vector where the value at
+                each index represents the mapping of the position of a single character
+                in the unannotated text to the position in the original text.
+            tokens (List[dict]): List of tokens represented as dictionaries. With "start"
+                indices referring to the unannotated text.
+        """
+        for i in range(len(tokens)):
+            tokens[i]["start"] = unann_to_ann_idx_map[tokens[i]["start"]]
+
+    def tokenize_using_spacy(self, text):
+        """Wrapper function used before tokenizing with Spacy. Combines all unannoted text spans
+        into a single string to pass to spacy for tokenization. Applies the correct offset to
+        the resulting tokens to align with the annotated text. This optimization reduces the overall
+        time needed for tokenization.
+
```

Comment:
Out of curiosity, what is restricting us from doing the same with other tokenizer types too? I mean, isn't this technique generic enough optimization across all types of tokenizers? Do you think this part of computation 
> Applies the correct offset to the resulting tokens to align with the annotated text.
offsets the computational speed we gain with combined text vs. multiple calls with each span?

Reply:
That's a good point, however, the reason the optimization works for Spacy is that Spacy guarantees that characters are not added or removed. Since the Text Preparation Pipeline has been designed to be extensible and fully customizable there isn't a guarantee that a user's tokenizer implementation will also uphold this property.